models/base/document.py

```diff
@@ -0,0 +1,231 @@
+        :param kwargs: value to each field
+        """
+        self._data = {}
+        for field_name, field in self._fields.items():
+            if field.primary_key and field_name not in kwargs:
+                raise self.PrimaryKeyNotSetError(field_name)
+            self.__setattr__(field_name, kwargs.get(field_name))
+        self._initialised = True
+        self._referenced_by: list[Document] = []
+
+    def save(self) -> None:
+        """
+        Recursively call save on the root-level document
+        """
+        for referrer in self._referenced_by:
+            referrer.save()
+
+    def delete(self) -> None:
+        """
+        Remove all references to this document and persist the change.
+        """
+        for referrer in list(self._referenced_by):  # copy with list() to avoid issues with removing items
+            referrer.__unlink_referee(self)
+            referrer.save()
+
+    @property
+    def key(self):
+        """
+        The primary key value of the document.
+        """
+        return getattr(self, self._primary_key)
+
+    def find_referred_by(self, referrer_type: type = None, field_name: str = None) -> Document:
+        """
+        Find the first document that references this document matching the criteria.
+        ReferrerNotFound exception is raised if no referrer matching the criteria is found.
+        :param referrer_type: optional type of the referrer as a criteria
+        :param field_name: optional field name of the referrer where the instance is referred as a criteria
+        :return: the first referrer matching the criteria
+        """
+        for referrer in self._referenced_by:
+            if referrer_type is not None and not isinstance(referrer, referrer_type):
+                continue
+            if field_name is not None and not self in getattr(referrer, field_name):
+                continue
+            return referrer
+        raise self.ReferrerNotFound(referrer_type, field_name)
+
+    def __eq__(self, other):
+        return self._data == other._data
+
+    def _add_referrer(self, referer):
+        if referer not in self._referenced_by:
+            self._referenced_by.append(referer)
+
+    def _remove_referrer(self, referer):
+        if referer in self._referenced_by:
+            self._referenced_by.remove(referer)
+
+    def __unlink_referee(self, referee):
+        for field_name, field in self._fields.items():
+            if isinstance(field, ReferenceDocumentsField):
+                documents = getattr(self, field_name)
+                if referee in documents:
+                    documents.remove(referee)
+        self.save()
+
+    def __str__(self):
+        return f'{self.__class__.__name__}({self._data})'
+
+
+class IndexedDocument(Document, metaclass=MetaIndexedDocument):
+    """
+    Base class for all root level documents, directly persisted to disk.
+    A primary key must be defined to index all active documents.
```

Comment:
Example for how to define this?

Reply:
I think you meant to comment on #27. 
I have added an example to the docstring.